### PR TITLE
Add strong typing to context, options, and paylaod in Functions codebase

### PR DESCRIPTION
**Pull request by inlined**
What it says in the box. The nebulous `context: any, options: any, payload: any` are now strongly typed.

To get part of this working I had to convert strings to Runtimes. I also removed a few lodash calls that I saw.

### CODE REVIEW

File: src/deploy/functions/release.ts [hunk 1]
```diff
@@ -12,22 +12,29 @@ import { promptForFunctionDeletion } from "./prompts";
...
   const fullDeployment = createDeploymentPlan(
-    payload.functions.byRegion,
-    context.existingFunctions,
+    payload.functions!.byRegion,
+
+    // Note: this is obviously a sketchy looking cast. And it's true; the shapes don't
+    // line up. But it just so happens that we don't hit any bugs with the current
+    // implementation of the function. This will all go away once everything uses
+    // backend.FunctionSpec.
+    (context.existingFunctions! as any) as deploymentPlanner.CloudFunctionTrigger[],
```
Comment (joehan) [inline src/deploy/functions/release.ts:37]: TIL that you can totally circumvent Typsecript by casting to any then to the type you want

File: src/deploy/functions/validate.ts [hunk 2]
```diff
@@ -15,11 +15,12 @@ const cjson = require("cjson");
 
 /**
  * Check that functions directory exists.
- * @param options options object.
+ * @param options options object. In prod is an args.Options; in tests can just be {cwd: string}
  * @param sourceDirName Relative path to source directory.
  * @throws { FirebaseError } Functions directory must exist.
  */
-export function functionsDirectoryExists(options: object, sourceDirName: string): void {
+export function functionsDirectoryExists(options: { cwd: string }, sourceDirName: string): void {
```
Comment (joehan) [inline src/deploy/functions/validate.ts:22]: Should this be `options: { cwd?: string, configPath?: string}`, like what resolveProjectPath takes? Its not immediately clear to me if this will only ever have cwd, or if it will have configPath depending on the options passed in.

**Review by joehan — COMMENTED** [review]
This looks like a really good direction to go in, and I really like the idea of moving things out of context into caching helper functions. Bryan and Sam may also be interested in this, they've been talking about trying to type context/options/payload across the whole codebase for a while now

File: src/deploy/functions/args.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    triggers: deploymentPlanner.CloudFunctionTrigger[];
+  };
+}
+
+// Options come from command-line options and stored config values
+// TODO: actually define all of this stuff in command.ts and import it from there.
+export interface Options {
+  cwd: string;
+  configPath: string;
+  // Why is there options.project and context.projectId?
```
[the remark this answers is not in the document]
Comment (inlined) [reply src/deploy/functions/args.ts:20]: So should I make sure I change code that refers to options.project to context.projectId?

File: src/deploy/functions/args.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    askWriteProjectFile(path: string, content: unknown): void;
+
+    projectDir: string;
+  };
+  filteredTargets: string[];
+  nonInteractive: boolean;
+  force: boolean;
+}
+
+export interface FunctionsSource {
```
[the remark this answers is not in the document]
Comment (inlined) [reply src/deploy/functions/args.ts:46]: Hmm... interesting point. In my work-in-progress branch I have `backend.existingBackend(context)`. Maybe I could do `prepareFunctionsUpload.functionsSource(context)`? I don't know that I want to have `functionsSource: prepareFunctionsUpload.FunctionsSource` because it'll create a weird dependency cycle.

How do you feel? (also, as an aside, I hate that we have file names named after a command, yet they act as a namespace for multiple things. Go-style namespacing ftw IMO)

File: src/deploy/functions/checkIam.ts [hunk 5]
```diff
@@ -51,15 +52,19 @@ export async function checkServiceAccountIam(projectId: string): Promise<void> {
  * @param options The command-wide options object.
  * @param payload The deploy payload.
  */
-export async function checkHttpIam(context: any, options: any, payload: any): Promise<void> {
-  const functionsInfo = payload.functions.triggers;
+export async function checkHttpIam(
```
[the remark this answers is not in the document]
Comment (inlined) [reply src/deploy/functions/checkIam.ts:55]: I have mixed feelings. I've gotten used to "context' being passed around in Go and I prefer editing Context as the only side-effect allowed in our API calls. I've got no love for "Options' though since the actual command line flags passed seem like too low level a concern for this code (e.g. options' project vs context's project). And I have yet to find the purpose of Payload.

File: src/deploy/functions/release.ts (hunk 1 above)
Comment (inlined) [reply src/deploy/functions/release.ts:37]: 🤫

File: src/deploy/functions/validate.ts (hunk 2 above)
Comment (inlined) [reply src/deploy/functions/validate.ts:22]: I'm a bit confused. Are you saying that `sourceDirName` should be part of the passed options? This is always an `args.Options`, which always has a `cwd`. I just changed it from `args.Options` because tests were annoying to fix if we required a full `args.Options`.

File: src/deploy/functions/args.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (joehan) [reply src/deploy/functions/args.ts:20]: Yeah, as far as I know, we pretty much should never use options.project outside of https://github.com/firebase/firebase-tools/blob/master/src/getProjectId.js

File: src/deploy/functions/validate.ts (hunk 2 above)
Comment (joehan) [reply src/deploy/functions/validate.ts:22]: Not quite - functionsDirectoryExists calls [resolveProjectPath](https://github.com/firebase/firebase-tools/blob/master/src/projectPath.ts#L11), which has the following signature: ```export function resolveProjectPath(
  options: { cwd?: string; configPath?: string },
  filePath: string
):``` 
I'm wondering if theres any edge cases where options.configPath would be defined here - from https://github.com/firebase/firebase-tools/blob/master/src/command.ts#L237, it looks like it comes from the --config flag.

File: src/deploy/functions/checkIam.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (joehan) [reply src/deploy/functions/checkIam.ts:55]: Agree with you on options and  (especially) payload.  We use this same "context, options, paylod" pattern for all of the deploy code paths, and payload might get used more in the other deploy modules (hosting, storage rules, etc), but its basically unused here. 

For context, I've bumped into a bunch of subtle bugs over my time with this codebase caused by using context as a mutable "global" state. I also find that it can make the code hard to follow sometimes - its not always obvious where context.blah was originally set or if it has been set at all. Having said that, I do think that your proposed changes (stronger typing, and moving complex objects like 'existingFunctions' out of context) alleviate these concerns a bit.

File: src/deploy/functions/args.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (inlined) [reply src/deploy/functions/args.ts]: Removed

File: src/deploy/functions/validate.ts (hunk 2 above)
Comment (inlined) [reply src/deploy/functions/validate.ts]: Documenting the configPath option. Making it optional since the existing tests don't use it, though it should always be present per my research into the Options type. Since both prod and test provide a CWD I'm leaving that as required.

**Review by joehan — APPROVED** [review]
